### ISSUE TITLE
Problem with segmentation result

### ISSUE THREAD

**Issue by LiXm1002** [VisualComputingInstitute/TrackR-CNN#30]
![WechatIMG1](https://user-images.githubusercontent.com/39939478/65473487-91986100-deaa-11e9-9fb0-d06c437401ea.jpeg)
As the picture show,the left of mask is straight when testing kitti dataset, and when I testing my own model,this situation also often appears,I'm sure the edge of my annotation is smooth,but the edge of testing result always straight. What should I do to avoid this happening?

**Comment by pvoigtlaender** [comment]
Our Mask R-CNN does not use a very high resolution, probably this causes the straight line. I think there is no easy fix.

**Comment by LiXm1002** [comment]
OK , Thanks~

**Comment by LiXm1002** [comment]
Oh,Thanks.

**Comment by LiXm1002** [comment]
Hi,@pvoigtlaender, do you have some good suggestions about this situation.  The segmentation result may affect processing later a lot. I want to make some modifications to it but I have no idea which part to start.

**Comment by pvoigtlaender** [comment]
You can try to increase the resolution of the mask head, but then you might not be able to reuse some if it's weights.
I think the 14 in this line https://github.com/VisualComputingInstitute/TrackR-CNN/blob/master/network/FasterRCNN.py#L316
is the resolution used for RoI align, i.e. 14x14. You can try to change it, but expect trouble. You probably only want to change the resolution used for the mask head and not the features for the other heads.

**Comment by LiXm1002** [comment]
Wow. Sounds great. I will try it. And what is this parameter ANCHOR_STRIDE = 12 （in datasets/util/Detection.py）using for? I want to try changing the value of it.

**Comment by pvoigtlaender** [comment]
I don't think ANCHOR_STRIDE will help you with better masks. I'd recommend not changing it unless you have a good reason to.

**Comment by LiXm1002** [comment]
You are right, change ANCHOR_STRIDE make it worse. I better try the method you said before.

**Comment by EchoAmor** [comment]
@LiXm1002 Sorry to disturb you , I'm new to tracking too ,and I run the code succefully but can't get results,can you tell me how you did the work?
#31

**Comment by LiXm1002** [comment]
I use "python main.py configs/conv3d_sep2 "{\"task\":\"forward_tracking\",\"dataset\":\"KITTI_segtrack_feed\",\"load_epoch_no\":5,\"batch_size\":5,\"export_detections\":true,\"do_tracking\":true,\"video_tags_to_load\":[\"0000\"]}""  and tracking result will be created in forwarded/tracking_data folder. Then use mots_tools-master/mots_vis/visualize_mots.py to make mask show on source images.

**Comment by LiXm1002** [comment]
How to use visualize_mots.py the author has written in README.md.

**Comment by EchoAmor** [comment]
@LiXm1002 
Thanks for replying!
The commands I used is the same as yours,but the results in /forwaded folder are 0 bytes,I screenshot the results in #31 , I didn't get any errors but the results is no data .
![mots1](https://user-images.githubusercontent.com/25685837/65745454-5017e800-e12e-11e9-9202-26650abe1a58.png)
![mots3](https://user-images.githubusercontent.com/25685837/65745462-57d78c80-e12e-11e9-9b69-b8dacfc6ebd9.png)

**Comment by LiXm1002** [comment]
@pvoigtlaender Hi.
In 316 line in network/FastRCNN.py `roi_resized = roi_align(rpn_input, boxes_on_featuremap, 14)` means output of roialign is 14*14, and  in 319 line `feature_fastrcnn_ = add_resnet_conv5(roi_resized, self.tower_setup)[0]` , is the shape of feature_fastrcnn_  7*7 or 14*14？

**Comment by LiXm1002** [comment]
When computing mask loss, both mask_logits and target_masks_for_fg are 14*14, and output of roialign is 14*14 too. but input of mask head should be 7*7, which step convert feature map 14*14 to 7*7?  It make me crazy.  T^T

**Comment by LiXm1002** [comment]
Oh,I got it, 'add_resnet_conv5' convert 14X14 to 7X7

**Comment by LiXm1002** [comment]
@pvoigtlaender Hi, I have a question, in 377 line of network/FasterRCNN `feature_fastrcnn_ = add_resnet_conv5(roi_resized, self.tower_setup)[0]` , the input of add_resnet_conv5 is 14X14 and output is 7X7, is there pooling is this part, or some other step make the feature smaller?

**Comment by ahnonay** [comment]
Yes, this is a normal resnet block with a stride of 2 in total

**Comment by LiXm1002** [comment]
Oh,I got it ,in first block of res_5 , a conv layer with filter size = 3X3,stride = 2, padding = VALID make the feature map 14X14 to 7X7.

**Comment by LiXm1002** [comment]
I tried design a parallel res_5 block for mask so that the input feature to mask branch is 14X14 and other branch is still 7X7 ,I want to try to see if this result can solve the problem that the mask becomes straight.But when training in epoch = 1，the loss is very very big abuot 7e+09 , and after dozens epoch loss become about 2000 , is this situation means there are  some problem? Should I let it continue training?

**Comment by pvoigtlaender** [comment]
It sounds like it's broken, I don't think letting it train will help, rather search for problems (might be hard to find).

**Comment by EchoAmor** [comment]
@LiXm1002 Hello，have u run the code on MOTS DATASET？ I get the tracking results but can not visualize and eval it , there are only kitti examples. If you have test on MOTS dataset ,can you teach me  how to visualize the results? thanks a lot!!!

**Comment by LiXm1002** [comment]
@EchoAmor Sorry,I didn't test MOT dataset before.

**Comment by LiXm1002** [comment]
@pvoigtlaender You're right. it's very very hard.  T_T

**Comment by EchoAmor** [comment]
Thanks anyway!

**Comment by LiXm1002** [comment]
@pvoigtlaender I want to change `resnet_conv5` part of code. Observing how this part becomes more complicated will affect the mask. And if I use pre-train model(trackrcnn_init/converted), loss decrease from 5.4664e+06. if not use pre-train model ,  loss is always about 3.2, it neither drops nor rises a lot. I just change code in `Resnet.py`. Do I need change other part such as gradient？ I don't know what is wrong?

**Comment by ahnonay** [comment]
You do not need to write any manual gradients. However, if you change the size of the conv5 output, you may need to adjust some other things as well. For example check if the mask loss still works in _create_losses (since the ground truth masks need to have another size as well now, I don't know right now if this works automatically).

**Comment by ahnonay** [comment]
Closing this for now, please reopen if needed
